File: java/business-context/src/main/java/cz/filipklimes/diploma/framework/businessContext/expression/numeric/LessOrEqualTo.java
package cz.filipklimes.diploma.framework.businessContext.expression.numeric;

import cz.filipklimes.diploma.framework.businessContext.weaver.BusinessOperationContext;
import cz.filipklimes.diploma.framework.businessContext.expression.BinaryOperator;
import cz.filipklimes.diploma.framework.businessContext.expression.Expression;

import java.math.BigDecimal;
import java.util.*;

public class LessOrEqualTo extends BinaryOperator<Boolean, BigDecimal, BigDecimal>
{

    public LessOrEqualTo(final Expression<BigDecimal> left, final Expression<BigDecimal> right)
    {
        super(left, right);
    }

    @Override
    public Boolean interpret(final BusinessOperationContext context)
    {
        return getLeft().interpret(context).compareTo(getRight().interpret(context)) <= 0;
    }

    @Override
    public Map<String, String> getProperties()
    {
        return Collections.emptyMap();
    }

    @Override
    public String getName()
    {
        return "numeric-lte";
    }

    @Override
    public String toString()
    {
        return String.format("%s <= %s", getLeft(), getRight());
    }

}

File: nodejs/business-context/src/expression/numeric/LessOrEqualTo.js
export default class LessOrEqualTo {
    constructor(left, right) {
        this.left = left
        this.right = right
    }

    interpret(context) {
        return this.left.interpret(context) <= this.right.interpret(context)
    }

    getName() {
        return 'numeric-lte'
    }

    getArguments() {
        return [this.left, this.right]
    }

    getProperties() {
        return {}
    }

    toString() {
        return this.left.toString() + " <= " + this.right.toString()
    }

    clone() {
        return new LessOrEqualTo(
            this.left.clone(),
            this.right.clone(),
        )
    }
}
